server/datastore/mysql/apple_mdm.go

```diff
@@ -4376,3 +4376,19 @@ AND TIMESTAMPDIFF(SECOND, h.detail_updated_at, NOW()) > ?;`)
 
 	return deviceUUIDs, nil
 }
+
+func (ds *Datastore) GetHostUUIDsWithPendingMDMAppleCommands(ctx context.Context) (uuids []string, err error) {
+	const stmt = `
+SELECT DISTINCT neq.id
+FROM nano_enrollment_queue neq
+LEFT JOIN nano_command_results ncr ON ncr.command_uuid = neq.command_uuid
+WHERE neq.active = 1 AND ncr.status IS NULL
```

Comment:
I wonder if that query with a DISTINCT could become a performance issue, though it's on the reader so it might not be as bad...

Reply:
hm good point, it does take considerable time in dogfood:

```
mysql> SELECT DISTINCT neq.id FROM nano_enrollment_queue neq LEFT JOIN nano_command_results ncr ON ncr.command_uuid = neq.command_uuid WHERE neq.active = 1 AND ncr.status IS NULL;

+--------------------------------------+
| id                                   |
+--------------------------------------+
REDACTED
+--------------------------------------+
19 rows in set (1 min 10.89 sec)

mysql> EXPLAIN SELECT DISTINCT neq.id FROM nano_enrollment_queue neq LEFT JOIN nano_command_results ncr ON ncr.command_uuid = neq.command_uuid WHERE neq.active = 1 AND ncr.status IS NULL;
+----+-------------+-------+------------+------+-------------------------------+--------------+---------+------------------------+---------+----------+-----------------------------------+
| id | select_type | table | partitions | type | possible_keys                 | key          | key_len | ref                    | rows    | filtered | Extra                             |
+----+-------------+-------+------------+------+-------------------------------+--------------+---------+------------------------+---------+----------+-----------------------------------+
|  1 | SIMPLE      | neq   | NULL       | ALL  | PRIMARY,command_uuid,priority | NULL         | NULL    | NULL                   | 1178015 |    10.00 | Using where; Using temporary      |
|  1 | SIMPLE      | ncr   | NULL       | ref  | command_uuid                  | command_uuid | 510     | fleet.neq.command_uuid |       8 |    10.00 | Using where; Not exists; Distinct |
+----+-------------+-------+------------+------+-------------------------------+--------------+---------+------------------------+---------+----------+-----------------------------------+
2 rows in set, 1 warning (0.20 sec)
```

A tiny improvement, but it's still at the filter stage would be to scope down and grab rows with `updated_at` in the last week or something like that.

Did you have another query in mind? I'm going to experiment with `NOT EXISTS`

Reply:
`DISTINCT` didn't made a difference, seems like the query is just bad with the indexes we have :(

adding a `neq.created_at >= NOW() - INTERVAL 7 DAY` made a huge impact (query took 1 second), so maybe combining that with a `LIMIT`? I'm thinking on doing both so hosts that are stuck don't clog the queue.

Reply:
:+1: makes sense to me! Yeah some hosts could be offline and you'd just keep try to send a push on every cron.